mcp_client.py

```diff
@@ -0,0 +1,152 @@
+import asyncio
+import json
+import os
+from contextlib import AsyncExitStack
+from typing import Dict, List, Optional, TypeAlias
+
+from mcp import ClientSession, StdioServerParameters
+from mcp.client.stdio import stdio_client
+
+from huggingface_hub import AsyncInferenceClient, ChatCompletionInputTool, ChatCompletionOutput
+from huggingface_hub.inference._providers import PROVIDER_T
+
+
+# Type alias for tool names
+ToolName: TypeAlias = str
+
+
+class MCPClient:
+    def __init__(
+        self,
+        *,
+        provider: PROVIDER_T,
+        model: str,
+        api_key: Optional[str] = None,
+    ):
+        self.client = AsyncInferenceClient(
+            provider=provider,
+            api_key=api_key,
+        )
+        self.model = model
+        # Initialize MCP sessions as a dictionary of ClientSession objects
+        self.sessions: Dict[ToolName, ClientSession] = {}
+        self.exit_stack = AsyncExitStack()
+        self.available_tools: List[ChatCompletionInputTool] = []
+
+    async def add_mcp_server(self, command: str, args: List[str], env: Dict[str, str]):
+        """Connect to an MCP server
+
+        Args:
+            todo
+        """
+        server_params = StdioServerParameters(command=command, args=args, env=env)
+
+        stdio_transport = await self.exit_stack.enter_async_context(stdio_client(server_params))
+        stdio, write = stdio_transport
+        session = await self.exit_stack.enter_async_context(ClientSession(stdio, write))
+
+        await session.initialize()
+
+        # List available tools
+        response = await session.list_tools()
+        tools = response.tools
+        print("\nConnected to server with tools:", [tool.name for tool in tools])
+
+        # Map tool names to their server for later lookup
+        for tool in tools:
+            self.sessions[tool.name] = session
+
+        self.available_tools += [
+            {
+                "type": "function",
+                "function": {
+                    "name": tool.name,
+                    "description": tool.description,
+                    "parameters": tool.inputSchema,
+                },
+            }
+            for tool in tools
+        ]
+
+    async def process_query(self, query: str) -> ChatCompletionOutput:
+        """Process a query using `self.model` and available tools"""
+        messages = [{"role": "user", "content": query}]
+
+        response = await self.client.chat.completions.create(
+            model=self.model,
+            messages=messages,
+            tools=self.available_tools,
+            tool_choice="auto",
+        )
+
+        # Process response and handle tool calls
+        tool_calls = response.choices[0].message.tool_calls
+        if tool_calls is None or len(tool_calls) == 0:
+            return response
+
+        for tool_call in tool_calls:
```

Comment:
Since we are in an async context, we should run the tools in coroutines in parallel 

(can be done in later PR)